### DISCUSSION TITLE
Questions about the difference in behavior between [Engine::flushAndWait] and [Engine::flush]

### DISCUSSION THREAD

**Discussion by mejai1206** [google/filament#6407] [Q&A] [answered]
While using [Engine::flushAndWait] and [Engine::flush] to receive the result of [readPixel], 
I had a question about the difference in operation between [Engine::flushAndWait] and [Engine::flush].

The situation is as follows.

I wrote the following code to receive the result of [readPixel].
```
void render() {
  if (renderer->beginFrame(window->getSwapChain())) {
    //......rendering..... (like: renderer->render(view))

    uint8_t* pixels = new uint8_t[outputSize];
    filament::backend::PixelBufferDescriptor buffer(
                    pixels,
                    outputSize,
                    backend::PixelBufferDescriptor::PixelDataFormat::RGBA,
                    backend::PixelBufferDescriptor::PixelDataType::UBYTE,
                    [](void* buffer, size_t size, void* user) {
                       // on callback..
                    }, nullptr);

    renderer->readPixels(0, 0, width, height, std::move(buffer));
    renderer->endFrame();
}
```

If I run this code,
it was confirmed that the callback that can receive the result is called asynchronously.

But if I draw only one frame,(run the render function only once) that callback never gets called.
(The program does not exit and is doing other things.)

In this situation, I was able to receive the result by using [Engine::flushAndWait] as shown in the code below (synchronously).
```
void render() {
  if (renderer->beginFrame(window->getSwapChain())) {
    //......rendering..... (like: renderer->render(view))

    uint8_t* pixels = new uint8_t[outputSize];
    filament::backend::PixelBufferDescriptor buffer(
                    pixels,
                    outputSize,
                    backend::PixelBufferDescriptor::PixelDataFormat::RGBA,
                    backend::PixelBufferDescriptor::PixelDataType::UBYTE,
                    [](void* buffer, size_t size, void* user) {
                       // on callback..
                    }, nullptr);

    renderer->readPixels(0, 0, width, height, std::move(buffer));
    renderer->endFrame();
    mEngine->flushAndWait(); //new added
}
```

After that, I wanted to receive the result asynchronously in this state, so I used [Engine::flush] like the code below.
```
void render() {
  if (renderer->beginFrame(window->getSwapChain())) {
    //......rendering..... (like: renderer->render(view))

    uint8_t* pixels = new uint8_t[outputSize];
    filament::backend::PixelBufferDescriptor buffer(
                    pixels,
                    outputSize,
                    backend::PixelBufferDescriptor::PixelDataFormat::RGBA,
                    backend::PixelBufferDescriptor::PixelDataType::UBYTE,
                    [](void* buffer, size_t size, void* user) {
                       // on callback..
                    }, nullptr);

    renderer->readPixels(0, 0, width, height, std::move(buffer));
    renderer->endFrame();
    mEngine->flush(); //new added
}
```

However, it was confirmed that the result could not be received when drawing only one frame.

What I'm curious about is the following:

As we can infer from the names of the functions,
I expected that the only difference between [Engine::flushAndWait] and [Engine::flush] is just callback timing.
But, there were more differences.(callback not working)
I'm wondering if this behavior is intended or if it's a bug.

Thank you!

**Comment by pixelflinger** [comment, marked as answer]
ah... very interesting question indeed!

`Engine::flushAndWait` is typically needed to write tests where you render one frame and exit.  You case is almost similar, so using `flushAndWait` works. But as you pointed out, it's not ideal because it blocks.

Calling `Engine.flush()` after `endFrame` in fact, has no effect because `endFrame` already does (something equivalent to) that. 

The problem, in fact, is how we're dispatching the callback. Because the *want* the callback to be dispatched on the main thread, we can only do that when you're calling some method in filament. Currently this is done during `beginFrame`. This is the real reason why you need another frame, and also why you never get the "last" callback, and therefore the reason why you need `flushAndWait`.

Okay now there is a solution. Filament allows you to specify a "handler" for your callback. In that handler, what *must* do is re-post the callback into your platform's message handling system (or build your own). e.g. on Android it would be the `Looper`.

You can specify a `handler` when you create the `PixelBufferDescriptor` (please see the documentation in the header).

Your handle will be called on a *service thread*, completely asynchronously. So you have to be *very careful* about threading issues and you must be very quick. As I said before, you should re-dispatch the callback using your platform's messaging system.

I hope this helps!

**Reply by mejai1206** [reply]
Thank you for your kind reply!
As you said, I can get post function asynchronously and re-dispatch our main message looper.

I have an additional question.

When the terminatation event is received, is there any way to waiting callback which is registered with handler on  last frame synchronously?

On exit, I want to safely consume all messages synchronously on our main thread.

Thank you!

**Reply by pixelflinger** [reply]
No there is not. You have to make sure to not destroy the `Engine` until all messages have been consumed.

You can do this in several ways. I would simply suggest to queue a custom "terminate" message to your handler and only destroy the `Engine` when you receive that message (assuming your handler processes messages in order).

**Reply by mejai1206** [reply]
Yes.
This answer really helped me a lot. thanks!
